driver-core/src/main/java/com/datastax/driver/core/RegularStatement.java

```diff
@@ -42,32 +70,1143 @@ protected RegularStatement() {}
+
+        ValueDefinition(String name, DataType type) {
+            this(-1, name, type);
+        }
+
+        private ValueDefinition(int index, String name, DataType type) {
+            this.index = index;
+            this.name = name;
+            this.type = type;
+        }
+
+        /**
+         * Return the index of the value in the statement.
+         * <p>
+         * This information is set if a positional setter (for example {@link #setDouble(int, double)}) was used to set the value.
+         * Note that the variables of a given statement are either all named or all positional.
+         *
+         * @return the index, or -1 if named variables were used on this statement.
+         */
+        public int getIndex() {
+            return index;
+        }
+
+        /**
+         * Return the name of the value in the statement.
+         * <p>
+         * This information is set if a named setter (for example {@link #setDouble(String, double)}) was used to set the value.
+         * Note that the variables of a given statement are either all named or all positional.
+         *
+         * @return the name, or {@code null} if positional variables were used on this statement.
+         */
+        public String getName() {
+            return name;
+        }
+
+        /**
+         * Return the CQL data type inferred for the value.
+         * <p>
+         * This information is determined from the method that was used to set the value. For example, if
+         * {@link #setDouble(String, double)} was used, it will be {@link DataType#cdouble()}. For non-standard types, it will
+         * be the CQL type of the first {@link CodecRegistry custom codec} that can handle the value.
+         * The only case where the CQL type can't be inferred is if the value was set with {@link #setBytesUnsafe(int, ByteBuffer)}
+         * or {@link #setToNull(int)}; then this method will return {@code null}.
+         *
+         * @return the data type, or {@code null} if this variable was set with {@link #setBytesUnsafe(int, ByteBuffer)}
+         * or {@link #setToNull(int)}.
+         */
+        public DataType getType() {
+            return type;
+        }
+
+        @Override
+        public String toString() {
+            return String.format("%s:%s",
+                (index >= 0) ? index : name,
+                (type == null) ? "<unknown type>" : type.toString());
+        }
+    }
+
+    // A value that was set on the statement.
+    private class Value {
+
+        private final ValueDefinition definition;
+
+        // The serialized form that will be sent to Cassandra alongside the query
+        private final ByteBuffer bytes;
+
+        private Value(ValueDefinition definition, ByteBuffer bytes) {
+            this.definition = definition;
+            this.bytes = bytes;
+        }
+
+        private <V> V as(TypeToken<V> javaType) {
+            if (OBJECT_TYPE.equals(javaType)) { // we're handling a getObject() call
```

Comment:
That's a good point, but my plan was to move OBJECT_TYPE to TypeTokens and make it public, which in theory allows users to pass it to get(). So it might be better to use `null` to indicate getObject (as you initially did).